pkg/util/ec2/ec2_tags.go

```diff
@@ -29,6 +30,46 @@ var (
 )
 
 func fetchEc2Tags(ctx context.Context) ([]string, error) {
+	if config.Datadog.GetBool("collect_ec2_tags_use_imds") {
+		// prefer to fetch tags from IMDS, falling back to the API
+		tags, err := fetchEc2TagsFromIMDS(ctx)
+		if err == nil {
+			return tags, nil
+		}
+
+		log.Debugf("Could not fetech tags from instance metadata (trying EC2 API instead): %s", err)
+	}
+
+	tags, err := fetchEc2TagsFromAPI(ctx)
+	if err == nil {
+		return tags, err
+	}
+
+	return nil, err
+}
+
+func fetchEc2TagsFromIMDS(ctx context.Context) ([]string, error) {
+	keysStr, err := getMetadataItem(ctx, "/tags/instance")
+	if err != nil {
+		return nil, err
+	}
+
+	// keysStr is a newline-separated list of strings containing tag keys
+	keys := strings.Split(keysStr, "\n")
+
+	tags := []string{}
+	for _, key := range keys {
```

Comment:
I believe the context does constrain that where necessary, although a few searches for indirect uses of this function show that it's usually context.TODO() (even in metadata collection).  Still, if consumers of this API want to limit the time, that's the mechanism.

The metadata fetch has a timeout, and if the first call times out then the others will not be run.  And otherwise, the IMDS is an on-host service so it's reliably fast.